src/libsync/syncengine.cpp

```diff
@@ -862,11 +860,24 @@ void SyncEngine::restoreOldFiles(SyncFileItemVector &syncItems)
             qCWarning(lcEngine) << "restoreOldFiles: RESTORING" << syncItem->_file;
             syncItem->_instruction = CSYNC_INSTRUCTION_CONFLICT;
             break;
-        case CSYNC_INSTRUCTION_REMOVE:
+            // Looks like the problem actually happens because of the server upgrade is done without enabling the 'maintenance'mode that would stop all the syncing
+            // until it's turned back off. Also, it looks like the files were removed from the server, and the "data-fingerprint" was updated as well (with a separate server command).
+            // This of course is one of those cases, when users would just need to do a full re-sync to fix the consequences of an incorrect upgrade procedure.
+            // But, the user I am talking to is refusing to do so, and saying that this is very hard to do as there's a lot of users and they are in different time-zones.
+            // I've been debugging this problem for a while. And here is what I propose:
+
+            // Based on my reasearch - the CSYNC_INSTRUCTION_REMOVE case makes no sense
+            // why do we want to re-upload a file that's been changed on the server? the remove instruction is only set if the file was actually removed
+            // for those cases when the user was modifying the file offline while it was removed from the server, the instruction is set to CSYNC_INSTRUCTION_NEW
+            // so the user's work won't be lost
+            // if it's a file that's been modified elsewhere (by other user - then we'll be having instruction CSYNC_INSTRUCTION_SYNC).
+            // Removing the case for CSYNC_INSTRUCTION_REMOVE would fix the issue with removed files being reuploaded because of the "data-fingerprint" has changed on the server
+            // while the files were removed just before or right after that.
+        /*case CSYNC_INSTRUCTION_REMOVE:
             qCWarning(lcEngine) << "restoreOldFiles: RESTORING" << syncItem->_file;
             syncItem->_instruction = CSYNC_INSTRUCTION_NEW;
             syncItem->_direction = SyncFileItem::Up;
-            break;
+            break;*/
```

Comment:
Indeed, I am still sure this would be the proper fix. According to unit-tests (that I've fixed in the subsequent commit), the file should be re-uploaded to a server, if it was removed before creating the backup - resulting in a backup not containing this file. But, if the file was removed before the backup - it means it should be removed no matter if a backup is performed later or if it is not performed. It is a normal file/folder removal that should be synced to all clients - hence it should be removed from all the clients and should go to a server trash bin like any normal removal. Re-uploading a file that was removed from the server makes no sense. Let's say, we had a folder "B1" and then we've removed it from the server (either via Web UI or by removing it on one of the clients, and then syncing the removal to a server"B1"). If the file was modified on the user's device while being offline, while the same file was also removed from the server at the same time, then, the next sync will just upload a file as a NEW file, instead of re-uploading the removed file.